### DISCUSSION TITLE
Getting TypeError: Cannot read properties of undefined (reading 'getFullText') when building

### DISCUSSION THREAD

**Discussion by jamezening** [egoist/tsup#909] [Q&A]
Hi,

I am using tsup to build a ts libary within a turborepo monorepo.

When I run `tsup` to build I am getting the error
`TypeError: Cannot read properties of undefined (reading 'getFullText')`

which is at ${path to node_modules}/tsup/dist/rollup.js:7415

```
const newProgram = createProgram$1(fileName, compilerOptions, tsconfig);
        programs.push(newProgram);
        const source = newProgram.getSourceFile(fileName); // this line
        return {
          code: source.getFullText(),
          source,
          program: newProgram
        };
```

The rollup version installed in the repo is `"rollup": "^3.2.5",`

Has anyone else had similar errors and maybe a solution?

Thank you.

**Comment by AnnaSakivska** [comment]
I faced the same issue. Did you find solution?

**Comment by kainbryanjones** [comment]
I realise this is quite an old issue now, but for anybody else encountering this I found what was the cause (at least for me).

**Circular Dependencies** may be the main cause here. In summary that means you have two modules (A and B) when Module A imports Module B and Module B imports Module A. When bundling this won't actually throw an error until you try to use the library.

I used [**dpmd**](https://www.npmjs.com/package/dpdm) to analyse circular dependencies in my code with the following command.

"detect-circular-deps": "dpdm --no-warning --no-tree --exit-code circular:1 src/index.ts"

I suggest putting "npm run detect-circular-deps" in your package.json "prebuild" and/or "prepack" scripts.

A common cause of circular dependencies is when using a barrel import system (exporting all modules within a directory from index.js / index.ts).
